smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+
+/**
+ * Index of member shape to associated resource property information.
+ */
+public final class PropertyBindingIndex implements KnowledgeIndex {
+    private final WeakReference<Model> model;
+    private final OperationIndex operationIndex;
+    private final Set<ShapeId> notPropertyMetaTraitSet;
+    private final Map<ShapeId, Boolean> memberShapeDoesNotRequireProperty = new HashMap<>();
+    private final Map<ShapeId, String> memberShapeToPropertyName = new HashMap<>();
+    private final Map<ShapeId, ShapeId> operationToInputPropertiesShape = new HashMap<>();
+    private final Map<ShapeId, ShapeId> operationToOutputPropertiesShape = new HashMap<>();
+
+    private PropertyBindingIndex(Model model) {
+        this.model = new WeakReference<>(model);
+        this.notPropertyMetaTraitSet = computeNotPropertyTraits();
+        this.operationIndex = OperationIndex.of(model);
+        IdentifierBindingIndex identifierIndex = IdentifierBindingIndex.of(model);
+
+        for (ResourceShape resourceShape : model.getResourceShapes()) {
+            Set<String> propertyNames = resourceShape.getProperties().keySet();
+            for (ShapeId operationShapeId : resourceShape.getAllOperations()) {
+                OperationShape operationShape = (OperationShape) model.getShape(operationShapeId).get();
+                Shape inputPropertiesShape = getInputPropertiesShape(operationShape);
+                operationToInputPropertiesShape.put(operationShapeId, inputPropertiesShape.getId());
+                for (MemberShape memberShape : inputPropertiesShape.members()) {
+                    if (identifierIndex.getOperationInputBindings(resourceShape, operationShape).values()
+                            .contains(memberShape.getMemberName())) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    } else {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(),
+                                doesNotRequireProperty(memberShape));
+                    }
+                    if (doesMemberShapeRequireProperty(memberShape.getId())
+                            || propertyNames.contains(memberShape.getMemberName())) {
+                        memberShapeToPropertyName.put(memberShape.getId(), getPropertyTraitName(memberShape)
+                            .orElse(memberShape.getMemberName()));
+                    }
+                }
+                // nesting is taking place, so index top level input/output members as not property.
+                if (!inputPropertiesShape.getId().equals(operationShape.getInputShape())) {
+                    for (MemberShape memberShape : model.expectShape(operationShape.getInputShape()).members()) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    }
+                }
+                Shape outputPropertiesShape = getOutputPropertiesShape(operationShape);
```

Comment:
Can these two really similar chunks of code be deduplicated? If not, can they be broken out to input and output methods or at least have a blank line to separate for reading?